### PR TITLE
HMAC verification codes in the database

### PR DESCRIPTION
**Pull request by sethvargo**
**Release Note**

```release-note
HMAC verification codes in the database - this prevents someone with access to the database from (easily) retrieving the plaintext tokens
```

### CODE REVIEW

File: pkg/database/token.go [hunk 1]
```diff
@@ -113,7 +113,12 @@ func (t *Token) Subject() *Subject {
 func (db *Database) ClaimToken(realmID uint, tokenID string, subject *Subject) error {
 	return db.db.Transaction(func(tx *gorm.DB) error {
 		var tok Token
-		if err := db.db.Set("gorm:query_option", "FOR UPDATE").Where("token_id = ? and realm_id = ?", tokenID, realmID).First(&tok).Error; err != nil {
+		if err := tx.
```
Comment (sethvargo) [inline pkg/database/token.go:116]: Note: this wasn't using the transaction before, fixed that.

File: pkg/database/token.go [hunk 2]
```diff
@@ -155,17 +160,29 @@ func (db *Database) VerifyCodeAndIssueToken(realmID uint, verCode string, accept
...
+		return nil, fmt.Errorf("failed to create hmac: %w", err)
+	}
+
 	var tok *Token
 	err = db.db.Transaction(func(tx *gorm.DB) error {
 		// Load the verification code - do quick expiry and claim checks.
 		// Also lock the row for update.
 		var vc VerificationCode
-		if err := db.db.Set("gorm:query_option", "FOR UPDATE").Where("(code = ? OR long_code = ?) AND realm_id = ?", verCode, verCode, realmID).First(&vc).Error; err != nil {
+		if err := tx.
```
Comment (sethvargo) [inline pkg/database/token.go:173]: Note this wasn't using the transaction before - fixed that.

File: pkg/database/token.go (hunk 1 above)
Comment (mikehelmick) [reply pkg/database/token.go:116]: thanks!

File: pkg/database/token.go [hunk 3]
```diff
@@ -195,14 +211,18 @@ func (db *Database) VerifyCodeAndIssueToken(realmID uint, verCode string, accept
...
 	})
+
 	return tok, err
 }
 
 func (db *Database) FindTokenByID(tokenID string) (*Token, error) {
 	var token Token
-	if err := db.db.Where("token_id = ?", tokenID).First(&token).Error; err != nil {
+	if err := db.db.
+		Or("token_id = ?", tokenID).
```
Comment (mikehelmick) [inline pkg/database/token.go:223]: Why did this change to Or?

Comment (sethvargo) [reply pkg/database/token.go:223]: Good catch, my mistake.

**Comment by sethvargo** [comment]
@mikehelmick fixed - PTAL

**Comment by sethvargo** [comment]
/retest

The failure seems to be related to our Twilio config - has something changed on CI @chaodaiG ?

**Comment by chaodaiG** [comment]
> /retest
> 
> The failure seems to be related to our Twilio config - has something changed on CI @chaodaiG?

Not that I know of. The prow job configs haven't been updated for 29 days https://github.com/GoogleCloudPlatform/oss-test-infra/tree/master/prow/prowjobs/google/exposure-notifications-verification-server. The build cluster hasn't been changed either

**Pull request merged**